Add documentation and development treats (#23)
decidim-vocdoni/decidim-module-vocdoni: README.md

# Decidim::Vocdoni

[![[CI] Lint](https://github.com/decidim-vocdoni/decidim-module-vocdoni/actions/workflows/lint.yml/badge.svg)](https://github.com/decidim-vocdoni/decidim-module-vocdoni/actions/workflows/lint.yml)
[![[CI] Tests](https://github.com/decidim-vocdoni/decidim-module-vocdoni/actions/workflows/test.yml/badge.svg)](https://github.com/decidim-vocdoni/decidim-module-vocdoni/actions/workflows/test.yml)
[![Maintainability](https://api.codeclimate.com/v1/badges/126b8ece66b8292802f3/maintainability)](https://codeclimate.com/github/decidim-vocdoni/decidim-module-vocdoni/maintainability)

:warning: This module is under development and is not ready to be used in production.

> An elections component for decidim's participatory spaces based on the [Vocdoni][https://vocdoni.app].

Reviewer: As we're moving from the Markdown reference links, we need to change the format too: 

![image](https://user-images.githubusercontent.com/717367/217754454-66c234b8-08fb-4938-8813-0a268e996c1e.png)

```suggestion
An elections component for decidim's participatory spaces based on the [Vocdoni](https://vocdoni.app).
```

(I'm a fan of reference links, but I can leave with the links duplicated too.)

Author: reference links tends to confuse me, but i can live with them too

```bash
bundle
bin/rails decidim_vocdoni:install:migrations

> bundle exec rails decidim_vocdoni:webpacker:install

Reviewer: Can we change this to be consistent? 

```suggestion
bin/rails decidim_vocdoni:webpacker:install
```

Author: done

order to create the dummy test app database.

Then to test how the module works in Decidim, start the development server:

```bash
$ DATABASE_USERNAME=<username> DATABASE_PASSWORD=<password> bin/rails s
```

Note that `bin/rails` is a convenient wrapper around the command `cd development_app; bundle exec rails`.

In case you are using [rbenv](https://github.com/rbenv/rbenv) and have the
[rbenv-vars](https://github.com/rbenv/rbenv-vars) plugin installed for it, you
can add the environment variables to the root directory of the project in a file
named `.rbenv-vars`. If these are defined for the environment, you can omit
defining these in the commands shown above.

#### Webpacker notes

As latest versions of Decidim, this repository uses Webpacker for Rails. This means that compilation
of assets is required every time a Javascript or CSS file is modified. Usually, this happens
automatically, but in some cases (specially when actively changes that type of files) you want to 
speed up the process. 

To do that, start in a separate terminal than the one with `bin/rails s`, and BEFORE it, the following command:

> 
> ```
> bin/webpack-dev-server

Reviewer: Missing markdown code language and also adding the dollar sign ($) for the prompt (for consistency). 

```suggestion

```bash
$ bin/webpack-dev-server
```

Author: I've added the bash indicator but removed the `$` symbol. It is not practical if you wan to do copy paste actions

```

#### Code Styling

Please follow the code styling defined by the different linters that ensure we
are all talking with the same language collaborating on the same project. This
project is set to follow the same rules that Decidim itself follows.

[Rubocop](https://rubocop.readthedocs.io/) linter is used for the Ruby language.

You can run the code styling checks by running the following commands from the
console:

> ```

Reviewer: Missing markdown code language. 

```suggestion
```bash
```

#### Non-Ruby Code Styling

There are other linters for Javascript and CSS. These run using NPM packages. You can
run the following commands:

1. `npm run lint`: Runs the linter for Javascript files.
2. `npm run lint-fix`: Automatically fix issues for Javascript files (if possible).
3. `npm run stylelint`: Runs the linter for SCSS files.
4. `npm run stylelint-fix`: Automatically fix issues for SCSS files (if possible).

### Testing

To run the tests run the following in the gem development path:

```bash
$ bundle
$ DATABASE_USERNAME=<username> DATABASE_PASSWORD=<password> bundle exec rake test_app
$ DATABASE_USERNAME=<username> DATABASE_PASSWORD=<password> bundle exec rspec
```

Note that the database user has to have rights to create and drop a database in
order to create the dummy test app database.

In case you are using [rbenv](https://github.com/rbenv/rbenv) and have the
[rbenv-vars](https://github.com/rbenv/rbenv-vars) plugin installed for it, you
can add these environment variables to the root directory of the project in a
file named `.rbenv-vars`. In this case, you can omit defining these in the
commands shown above.

### Test code coverage

> Code coverage report is generated automatically in a folder named `coverage` in the project root which contains
> the code coverage report.
> 
> ```bash
> firefox coverage/index.html
> ```

Reviewer: We should explicilty tell that's generated after the test suite is run. 

Also, I prefer not to assume that everyone is using firefox (even though they should 😄).

Can we change this to something like this 🙏🏽 ? 

```suggestion
Code coverage report is generated automatically after running the test suite, in a folder
named `coverage` in the project root which contains the code coverage report. It's
available at the file ./coverage/index.html. If you're using GNU/Linux, you can open
it with `xdg-open ./coverage/index.html`.
```

Author: That was just me being political advocate! :smile:

Reviewer: Ok, I'll let it pass as it's Firefox 😅

### Localization

If you would like to see this module in your own language, you can help with its
translation at Crowdin:

https://crowdin.com/project/decidim-module-vocdoni

## License

This engine is distributed under the GNU AFFERO GENERAL PUBLIC LICENSE.
See [LICENSE-AGPLv3.txt](LICENSE-AGPLv3.txt).

> As this module works with the Vocdoni App, see https://vocdoni.app/ for more information about their open source licenses.

Reviewer: Just a couple corrections regarding this line:

1. We're actually using the Vocdoni API/SDK, so we should link to https://vocdoni.io/  (I corrected that in other PR) 
2. Regarding their license, it's not relevant for us, as using an API it's not considered a derivative work. See https://opensource.stackexchange.com/a/8204 (They're using the AGPL for their SDK https://github.com/vocdoni/vocdoni-sdk/blob/main/LICENSE)

Author: Ok, I guess changing the link is enough?

Reviewer: Mmm yes, let's change the link and the name too

> As this module works with the Vocdoni App, see https://vocdoni.app/ for more information about their open source licenses.

Reviewer: ```suggestion
As this module works with the Vocdoni SDK, see https://vocdoni.io/ for more information about their open source licenses.
```

Author: done
